lib/net.js

```diff
@@ -400,6 +400,12 @@ Socket.prototype.setTimeout = function(msecs, callback) {
 
 
 Socket.prototype._onTimeout = function() {
+  const writeQueueSize = this._handle.writeQueueSize;
+  if (writeQueueSize > 0 &&
+      writeQueueSize !== this._handle.getWriteQueueSize()) {
+    this._unrefTimer();
+    return;
```

Comment:
I don't understand why the 'timeout' event is suppressed here.

As well, can you add a comment explaining the checks?  They're going to look tautologically wrong to a casual reader ("when would the queue size not equal the queue size?")

Reply:
So the current bug this is dealing with is that once you had something off to the backing stream, it can be actively writing but the socket doesn't know about it and if the timeout happens, it'll just destroy itself and the write that's currently occurring.

We don't update `writeQueueSize` as the write is ongoing, we only set it at the very beginning. So when one uses `this._handle.getWriteQueueSize()` they get an updated value and compare it to what it was when the write started. If the value changed the timeout is wrong because we're still writing behind the scenes, if the value didn't change then this is likely an actual timeout.

Does that sort of make some sense?

Reply:
`this._handle.getWriteQueueSize` shouldn't run unless prevQueueSize > 0 hence the current format. (It's not terribly slow but we also don't want to run it on every single timeout.) I'll adjust some things & add comment to make it clearer what's going on.

Re: timeout reset, it does. When the write is done, it triggers 'Socket.prototype.afterWrite` which calls `_unrefTimer`.